pennylane/_qutrit_device.py

```diff
@@ -0,0 +1,383 @@
+        raise NotImplementedError
+
+    def generate_samples(self):
+        r"""Returns the computational basis samples generated for all wires.
+
+        Note that PennyLane uses the convention :math:`|q_0,q_1,\dots,q_{N-1}\rangle` where
+        :math:`q_0` is the most significant trit.
+
+        .. warning::
+
+            This method should be overwritten on devices that
+            generate their own computational basis samples, with the resulting
+            computational basis samples stored as ``self._samples``.
+
+        Returns:
+             array[complex]: array of samples in the shape ``(dev.shots, dev.num_wires)``
+        """
+        number_of_states = 3**self.num_wires
+
+        rotated_prob = self.analytic_probability()
+
+        samples = self.sample_basis_states(number_of_states, rotated_prob)
+        return QutritDevice.states_to_ternary(samples, self.num_wires)
+
+    @staticmethod
+    def generate_basis_states(num_wires, dtype=np.uint32):
+        """
+        Generates basis states in ternary representation according to the number
+        of wires specified.
+
+        Args:
+            num_wires (int): the number wires
+            dtype=np.uint32 (type): the data type of the arrays to use
+
+        Returns:
+            array[int]: the sampled basis states
+        """
+        basis_states_generator = itertools.product((0, 1, 2), repeat=num_wires)
+        return np.fromiter(itertools.chain(*basis_states_generator), dtype=dtype).reshape(
+            -1, num_wires
+        )
+
+    @staticmethod
+    def states_to_ternary(samples, num_wires, dtype=np.int64):
+        """Convert basis states from base 10 to ternary representation.
+
+        This is an auxiliary method to the generate_samples method.
+
+        Args:
+            samples (array[int]): samples of basis states in base 10 representation
+            num_wires (int): the number of qutrits
+            dtype (type): Type of the internal integer array to be used. Can be
+                important to specify for large systems for memory allocation
+                purposes.
+
+        Returns:
+            array[int]: basis states in ternary representation
+        """
+        ternary_arr = []
+        for sample in samples:
+            num = []
+            for _ in range(num_wires):
+                sample, r = divmod(sample, 3)
+                num.append(r)
+
+            ternary_arr.append(num[::-1])
+
+        return np.array(ternary_arr, dtype=dtype)
+
+    def density_matrix(self, wires):
+        """Returns the reduced density matrix prior to measurement.
+
+        .. note::
+
+            Only state vector simulators support this property. Please see the
+            plugin documentation for more details.
+        """
+        # TODO: Add density matrix support. Currently, qml.math is hard-coded to work only with qubit states,
```

Comment:
I completely agree. That being said, I think that while I'm working on qutrits and not on qudits, I think it's best to limit the changes I make to new files rather than altering already existing ones, and `default.qutrit` will have its own internal method for returning the density matrix which doesn't rely on the method in `qml.math`. The other measurement types that might require changes to conform to arbitrary dimensionality are `VnEntropy` and `MutualInfo`.

Reply:
Agreed - best left as a "for later" task, after all the core functionality is implemented :+1: